File: web/src/api/modules/cicd.ts
import apiService from '../api';
import type { ApiResponse, PaginatedResponse } from '../api';

export type TriggerMode = 'manual' | 'source-event' | 'both';
export type TriggerType = 'manual' | 'source-event';
export type ReleaseStatus = 'pending_approval' | 'approved' | 'rejected' | 'executing' | 'succeeded' | 'failed' | 'rolled_back';

export interface ServiceCIConfig {
  id: number;
  service_id: number;
  repo_url: string;
  branch: string;
  build_steps: string[];
  artifact_target: string;
  trigger_mode: TriggerMode;
  status: string;
  updated_by: number;
  created_at: string;
  updated_at: string;
}

export interface CIRunRecord {
  id: number;
  service_id: number;
  ci_config_id: number;
  trigger_type: TriggerType;
  status: string;
  reason: string;
  triggered_by: number;
  triggered_at: string;
  created_at: string;
}

export interface DeploymentCDConfig {
  id: number;
  deployment_id: number;
  env: string;
  strategy: 'rolling' | 'blue-green' | 'canary';
  strategy_config: Record<string, any>;
  approval_required: boolean;
  updated_by: number;
  created_at: string;
  updated_at: string;
}

export interface ReleaseRecord {
  id: number;
  service_id: number;
  deployment_id: number;
  env: string;
  version: string;
  strategy: string;
  status: ReleaseStatus;
  triggered_by: number;
  approved_by: number;
  approval_comment: string;
  rollback_from_release_id: number;
  started_at?: string;
  finished_at?: string;
  created_at: string;
  updated_at: string;
}

export interface ReleaseApproval {
  id: number;
  release_id: number;
  approver_id: number;
  decision: 'approved' | 'rejected';
  comment: string;
  created_at: string;
}

export interface TimelineEvent {
  id: number;
  service_id: number;
  deployment_id: number;
  release_id: number;
  event_type: string;
  actor_id: number;
  payload: any;
  created_at: string;
}

export const cicdApi = {
  getServiceCIConfig(serviceId: number): Promise<ApiResponse<ServiceCIConfig>> {
    return apiService.get(`/cicd/services/${serviceId}/ci-config`);
  },

  putServiceCIConfig(serviceId: number, payload: {
    repo_url: string;
    branch?: string;
    build_steps?: string[];
    artifact_target: string;
    trigger_mode: TriggerMode;
  }): Promise<ApiResponse<ServiceCIConfig>> {
    return apiService.put(`/cicd/services/${serviceId}/ci-config`, payload);
  },

  deleteServiceCIConfig(serviceId: number): Promise<ApiResponse<void>> {
    return apiService.delete(`/cicd/services/${serviceId}/ci-config`);
  },

  triggerCIRun(serviceId: number, payload: { trigger_type: TriggerType; reason?: string }): Promise<ApiResponse<CIRunRecord>> {
    return apiService.post(`/cicd/services/${serviceId}/ci-runs/trigger`, payload);
  },

  listCIRuns(serviceId: number): Promise<ApiResponse<PaginatedResponse<CIRunRecord>>> {
    return apiService.get(`/cicd/services/${serviceId}/ci-runs`);
  },

  getDeploymentCDConfig(deploymentId: number, env?: string): Promise<ApiResponse<DeploymentCDConfig>> {
    return apiService.get(`/cicd/deployments/${deploymentId}/cd-config`, { params: { env } });
  },

  putDeploymentCDConfig(deploymentId: number, payload: {
    env: string;
    strategy: 'rolling' | 'blue-green' | 'canary';
    strategy_config?: Record<string, any>;
    approval_required?: boolean;
  }): Promise<ApiResponse<DeploymentCDConfig>> {
    return apiService.put(`/cicd/deployments/${deploymentId}/cd-config`, payload);
  },

  triggerRelease(payload: {
    service_id: number;
    deployment_id: number;
    env: string;
    version: string;
  }): Promise<ApiResponse<ReleaseRecord>> {
    return apiService.post('/cicd/releases', payload);
  },

  listReleases(params?: { service_id?: number; deployment_id?: number }): Promise<ApiResponse<PaginatedResponse<ReleaseRecord>>> {
    return apiService.get('/cicd/releases', { params });
  },

  approveRelease(id: number, comment?: string): Promise<ApiResponse<ReleaseRecord>> {
    return apiService.post(`/cicd/releases/${id}/approve`, { comment: comment || '' });
  },

  rejectRelease(id: number, comment?: string): Promise<ApiResponse<ReleaseRecord>> {
    return apiService.post(`/cicd/releases/${id}/reject`, { comment: comment || '' });
  },

  rollbackRelease(id: number, payload: { target_version: string; comment?: string }): Promise<ApiResponse<ReleaseRecord>> {
    return apiService.post(`/cicd/releases/${id}/rollback`, payload);
  },

  listApprovals(releaseId: number): Promise<ApiResponse<PaginatedResponse<ReleaseApproval>>> {
    return apiService.get(`/cicd/releases/${releaseId}/approvals`);
  },

  getServiceTimeline(serviceId: number): Promise<ApiResponse<PaginatedResponse<TimelineEvent>>> {
    return apiService.get(`/cicd/services/${serviceId}/timeline`);
  },
};

File: api/cicd/v1/cicd.go
package v1

import "time"

type UpsertServiceCIConfigReq struct {
	RepoURL        string   `json:"repo_url" binding:"required"`
	Branch         string   `json:"branch"`
	BuildSteps     []string `json:"build_steps"`
	ArtifactTarget string   `json:"artifact_target" binding:"required"`
	TriggerMode    string   `json:"trigger_mode" binding:"required"` // manual|source-event|both
}

type ServiceCIConfigResp struct {
	ID             uint      `json:"id"`
	ServiceID      uint      `json:"service_id"`
	RepoURL        string    `json:"repo_url"`
	Branch         string    `json:"branch"`
	BuildSteps     []string  `json:"build_steps"`
	ArtifactTarget string    `json:"artifact_target"`
	TriggerMode    string    `json:"trigger_mode"`
	Status         string    `json:"status"`
	UpdatedBy      uint      `json:"updated_by"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}

type TriggerCIRunReq struct {
	TriggerType string `json:"trigger_type" binding:"required"` // manual|source-event
	Reason      string `json:"reason"`
}

type CIRunResp struct {
	ID          uint      `json:"id"`
	ServiceID   uint      `json:"service_id"`
	CIConfigID  uint      `json:"ci_config_id"`
	TriggerType string    `json:"trigger_type"`
	Status      string    `json:"status"`
	Reason      string    `json:"reason"`
	TriggeredBy uint      `json:"triggered_by"`
	TriggeredAt time.Time `json:"triggered_at"`
	CreatedAt   time.Time `json:"created_at"`
}

type UpsertDeploymentCDConfigReq struct {
	Env              string         `json:"env" binding:"required"`
	RuntimeType      string         `json:"runtime_type"`
	Strategy         string         `json:"strategy" binding:"required"` // rolling|blue-green|canary
	StrategyConfig   map[string]any `json:"strategy_config"`
	ApprovalRequired bool           `json:"approval_required"`
}

type DeploymentCDConfigResp struct {
	ID               uint           `json:"id"`
	DeploymentID     uint           `json:"deployment_id"`
	Env              string         `json:"env"`
	RuntimeType      string         `json:"runtime_type"`
	Strategy         string         `json:"strategy"`
	StrategyConfig   map[string]any `json:"strategy_config"`
	ApprovalRequired bool           `json:"approval_required"`
	UpdatedBy        uint           `json:"updated_by"`
	CreatedAt        time.Time      `json:"created_at"`
	UpdatedAt        time.Time      `json:"updated_at"`
}

type TriggerReleaseReq struct {
	ServiceID    uint   `json:"service_id" binding:"required"`
	DeploymentID uint   `json:"deployment_id" binding:"required"`
	Env          string `json:"env" binding:"required"`
	RuntimeType  string `json:"runtime_type"`
	Version      string `json:"version" binding:"required"`
}

type ReleaseDecisionReq struct {
	Comment string `json:"comment"`
}

type RollbackReleaseReq struct {
	TargetVersion string `json:"target_version" binding:"required"`
	Comment       string `json:"comment"`
}

type ReleaseResp struct {
	ID                    uint       `json:"id"`
	ServiceID             uint       `json:"service_id"`
	DeploymentID          uint       `json:"deployment_id"`
	Env                   string     `json:"env"`
	RuntimeType           string     `json:"runtime_type"`
	Version               string     `json:"version"`
	Strategy              string     `json:"strategy"`
	Status                string     `json:"status"`
	TriggeredBy           uint       `json:"triggered_by"`
	ApprovedBy            uint       `json:"approved_by"`
	ApprovalComment       string     `json:"approval_comment"`
	RollbackFromReleaseID uint       `json:"rollback_from_release_id"`
	Diagnostics           any        `json:"diagnostics"`
	StartedAt             *time.Time `json:"started_at,omitempty"`
	FinishedAt            *time.Time `json:"finished_at,omitempty"`
	CreatedAt             time.Time  `json:"created_at"`
	UpdatedAt             time.Time  `json:"updated_at"`
}

type ReleaseApprovalResp struct {
	ID         uint      `json:"id"`
	ReleaseID  uint      `json:"release_id"`
	ApproverID uint      `json:"approver_id"`
	Decision   string    `json:"decision"`
	Comment    string    `json:"comment"`
	CreatedAt  time.Time `json:"created_at"`
}

type ReleaseTimelineEventResp struct {
	ID           uint      `json:"id"`
	ServiceID    uint      `json:"service_id"`
	DeploymentID uint      `json:"deployment_id"`
	ReleaseID    uint      `json:"release_id"`
	EventType    string    `json:"event_type"`
	ActorID      uint      `json:"actor_id"`
	Payload      any       `json:"payload"`
	CreatedAt    time.Time `json:"created_at"`
}
